src/statics.rs

```diff
@@ -14,60 +19,213 @@ once_cell::sync::Lazy::new(|| {
+            let mut next_pos = good_pos + 2;
+            while next_pos != end_pos {
+                if events_slice[next_pos].is_null() {
+                    break;
+                }
+                if events_slice[good_pos] != events_slice[next_pos] {
+                    good_pos += 1;
+                    events_slice[good_pos] = events_slice[next_pos];
+                }
+                next_pos += 1;
+            }
+            break;
+        }
+        if events_slice[good_pos + 1].is_null() {
+            break;
         }
+        good_pos += 1;
+    }
 
-        // Sort spurious nulls to the end
-        events_slice.sort_unstable_by(|a, b| b.cmp(a));
-
-        // Remove spurious duplicates
-        let end_pos = events_slice.len();
-        let mut good_pos = 0;
-        while good_pos != end_pos - 1 {
-            if events_slice[good_pos] == events_slice[good_pos + 1] {
-                let mut next_pos = good_pos + 2;
-                while next_pos != end_pos {
-                    if events_slice[good_pos] != events_slice[next_pos] {
-                        good_pos += 1;
-                        events_slice[good_pos] = events_slice[next_pos];
+    // Explicitly set all the values at the end to null
+    let mut next_pos = good_pos + 1;
+    while next_pos != end_pos {
+        events_slice[next_pos] = core::ptr::null();
+        next_pos += 1;
+    }
+
+    let bh = FnvHasher::default();
+
+    let mut map: Box<[core::mem::MaybeUninit<ParsedEventMetadata>]> = Box::new_uninit_slice(good_pos + 1);
+    next_pos = 0;
+    while next_pos <= good_pos {
+        // SAFETY The above code as already validated that events_slice[0..good_pos] are non-null pointers
+        let next = unsafe { &*events_slice[next_pos] };
+        let identity_hash = bh.hash_one(&next.identity);
+        // SAFETY The pointer being written to is valid (we just allocated it above) and aligned (it was allocated with the same type as being written)
+        unsafe { map[next_pos].as_mut_ptr().write(ParsedEventMetadata { identity_hash, meta: next }) };
+        next_pos += 1;
+    }
+    // SAFETY We've explicitly initialized all the values now
+    let mut sorted = unsafe { map.assume_init() };
+    sorted.sort_unstable_by(|a, b| b.cmp(a));
+    sorted
+});
+
+impl core::cmp::PartialEq for ParsedEventMetadata {
+    fn eq(&self, other: &Self) -> bool {
+        cmp::Ordering::Equal == self.cmp(other)
+    }
+}
+
+impl core::cmp::Eq for ParsedEventMetadata {}
+
+impl core::cmp::PartialOrd for ParsedEventMetadata {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl core::cmp::Ord for ParsedEventMetadata {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.identity_hash.cmp(&other.identity_hash) {
+            cmp::Ordering::Equal => {
+                match self.meta.identity == other.meta.identity {

```

Comment:
This makes me nervous. You're using one rule for equality and a different rule for GT/LT. That can get into bad behavior.

Do we use GT/LT for anything other than sorting/probing the map?

If so, you should be comparing ONLY the hash. Two ParsedEventMetadata items with equal hashes will compare equal, even if the identities are not equal. When you look them up, you'll hit the lower bound of them and perform identity == for each one to get actual equality, i.e. you'll implement standard hashtable functionality but with less allocation.